### PR TITLE
SONARIAC-789 Take dynamic blocks into account when detecting absence of properties

### PR DESCRIPTION
**Pull request by petertrr**

### CODE REVIEW

File: iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/tree/impl/BlockTreeImpl.java [hunk 1]
```diff
@@ -30,15 +30,20 @@
...
   private final BodyTree body;
-  private Kind kind;
+  private final Kind kind;
 
-  public BlockTreeImpl(SyntaxToken key, @Nullable List<LabelTree> labels, BodyTree body, Kind kind) {
+  public BlockTreeImpl(@Nullable SyntaxToken dynamic, SyntaxToken key, @Nullable List<LabelTree> labels, BodyTree body, Kind kind) {
+    this.dynamicKeyword = dynamic;
     this.key = key;
-    this.labels = labels != null ? labels : Collections.emptyList();
+    this.labels = requireNonNullElse(labels, Collections.emptyList());
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/tree/impl/BlockTreeImpl.java:46]: Wasn't aware this was existing, I like it, will keep this in mind!

File: iac-extensions/terraform/src/test/java/org/sonar/iac/terraform/TestTreeBuilders.java [hunk 2]
```diff
@@ -57,7 +58,7 @@ private SyntaxTokenBuilder() {
     }
 
     public static SyntaxToken token(String value) {
-      return new SyntaxTokenImpl(value, null, Collections.emptyList());
+      return new SyntaxTokenImpl(value, range(1, 0, value), Collections.emptyList());
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/terraform/src/test/java/org/sonar/iac/terraform/TestTreeBuilders.java:61]: Not sure why we replaced the null range here?

File: iac-extensions/terraform/src/test/java/org/sonar/iac/terraform/parser/BlockTest.java [hunk 3]
```diff
@@ -20,26 +20,37 @@
...
       .matches("a { /* a */  /* b */\n}")
+      .matches("""
+        dynamic "a" "label" {
+        }""")
+      .matches("""
+        dynamic a {
+        }""")
+      .matches("""
+        dynamic a "label" {
+        }""")
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/terraform/src/test/java/org/sonar/iac/terraform/parser/BlockTest.java:46]: Nitpick: for completeness `dynamic "a" label { }`

File: iac-extensions/terraform/src/test/java/org/sonar/iac/terraform/parser/BlockTest.java [hunk 4]
```diff
@@ -49,8 +60,32 @@ void test_oneLineBlock() {
...
+        }
+      }
+      """;
+
+    Assertions.assertThat(HclLexicalGrammar.BLOCK)
+      .extracting(parser -> parser.parse(code))
+      .isNotNull()
+      .satisfies(tree -> {
+        var blockTree = (BlockTree) tree;
+        org.assertj.core.api.Assertions.assertThat(blockTree.isDynamic()).isTrue();
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/terraform/src/test/java/org/sonar/iac/terraform/parser/BlockTest.java:86]: I'm not sure if I saw it, but did you assert that `isDynamic` return `false` in another test?

**Review by rudy-regazzoni-sonarsource — APPROVED** [review]
LGTM! Very nice job! Please just see my few comments.

File: iac-extensions/terraform/src/test/java/org/sonar/iac/terraform/TestTreeBuilders.java (hunk 2 above)
Comment (petertrr) [reply iac-extensions/terraform/src/test/java/org/sonar/iac/terraform/TestTreeBuilders.java:61]: It was in an earlier attempt on this PR, where something was relying on this range. As I removed that code, I'll rollback this change as well.

**Pull request merged**